diff --git a/bbl/bbl.lds b/bbl/bbl.lds
index 9d51f1d..26d8b8c 100644
--- a/bbl/bbl.lds
+++ b/bbl/bbl.lds
@@ -42,14 +42,10 @@ SECTIONS
   _etext = .;
 
   /*--------------------------------------------------------------------*/
-  /* HTIF, isolated onto separate page                                  */
+  /* HTIF                                                               */
   /*--------------------------------------------------------------------*/
-  .htif :
-  {
-    PROVIDE( __htif_base = . );
-    *(.htif)
-  }
-  . = ALIGN(0x1000);
+  tohost = 0x40008000;
+  fromhost = 0x40008008;
 
   /*--------------------------------------------------------------------*/
   /* Initialized data segment                                           */
diff --git a/machine/htif.c b/machine/htif.c
index 7d2265c..4480c03 100644
--- a/machine/htif.c
+++ b/machine/htif.c
@@ -8,18 +8,12 @@
 #include <string.h>
 
 extern uint64_t __htif_base;
-volatile uint64_t tohost __attribute__((section(".htif")));
-volatile uint64_t fromhost __attribute__((section(".htif")));
+volatile extern uint64_t tohost;
+volatile extern uint64_t fromhost;
 volatile int htif_console_buf;
 static spinlock_t htif_lock = SPINLOCK_INIT;
 uintptr_t htif;
 
-#define TOHOST(base_int)	(uint64_t *)(base_int + TOHOST_OFFSET)
-#define FROMHOST(base_int)	(uint64_t *)(base_int + FROMHOST_OFFSET)
-
-#define TOHOST_OFFSET		((uintptr_t)tohost - (uintptr_t)__htif_base)
-#define FROMHOST_OFFSET		((uintptr_t)fromhost - (uintptr_t)__htif_base)
-
 static void __check_fromhost()
 {
   uint64_t fh = fromhost;
diff --git a/pk/pk.lds b/pk/pk.lds
index 75af7c5..eca4564 100644
--- a/pk/pk.lds
+++ b/pk/pk.lds
@@ -42,14 +42,10 @@ SECTIONS
   _etext = .;
 
   /*--------------------------------------------------------------------*/
-  /* HTIF, isolated onto separate page                                  */
+  /* HTIF                                                               */
   /*--------------------------------------------------------------------*/
-  .htif :
-  {
-    PROVIDE( __htif_base = . );
-    *(.htif)
-  }
-  . = ALIGN(0x1000);
+  tohost = 0x40008000;
+  fromhost = 0x40008008;
 
   /*--------------------------------------------------------------------*/
   /* Initialized data segment                                           */
